Add `generate_opcheck_tests`, a PT2 crossref testing mechanism

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106903

This PR adds `generate_opcheck_tests`. This is a utility that adds
additional crossref tests to an existing TestCase that has tests that
invokes operators. The main use case is if you have a large test suite
that already exercises operators and want to add automated testing that
the operators are correct, without actually refactoring your code into
something like OpInfos.

Given a `test_` method of a TestCase, we will generate one new
additional test for each of {schema correctness, autograd registration,
faketensor rule, aot_autograd static shapes, aot_autograd dynamic
shapes}. Each newly generated test runs the original test method under a
special torch_function mode (OpCheckMode) that intercepts
`op(*args, **kwargs)` calls and additional passes (op, args, kwargs) to
a separate function (e.g. SchemaCheck).

Nitty-gritty details:
- If a test is named test_cumsum, we end up generating new tests
(`test_schema__test_cumsum`, `test_<something>__test_cumsum`)
- Users can provide a dictionary of expected failures / skips  that is indexed on
operators. This gives us a sense of what operators support PT2 and which
operators require fixing before they support PT2.

Due to some co-dev limitations, I'm planning on landing this PR first
and then using it to add crossref testing for internal tests and
fbgemms. I could squash this PR with the internal changes if we want to
see how that works out, just let me know.

Test Plan:
- We create a mini op test suite called MiniOpTests.
- Then, we use `generate_opcheck_tests` to generate tests onto it.
- We have our own test xfail list to check that the things that should
fail do fail.
- Finally, there is a separate TestGenerateOpcheckTests that checks that
the correct number of tests were generated and also tests some helper
functions.

zou3519 committed:
Add `generate_opcheck_tests`, a PT2 crossref testing mechanism

This PR adds `generate_opcheck_tests`. Given a TestCase, it will for all
existing tests, use the existing test to generate additional tests for:
- schema correctness
- autograd registration
- faketensor rule
- aot_autograd static shapes
- aot_autograd dynamic shapes

This set of tests check that an operator works with the cross product of
features used in PT2.

Concretely:
- If a test is named test_cumsum, we end up generating new tests
(`test_schema__test_cumsum`, `test_<something>__test_cumsum`)
- The new test runs the original test, captures calls to custom operators,
and then uses the calls to the custom operators to run the correct test. In
PyTorch core we call this style of testing "crossref testing", hence the
title.
- Users can provide a dictionary of expected failures / skips  that is indexed on
operators. This gives us a sense of what operators support PT2 and which
operators require fixing before they support PT2.

`generate_opcheck_tests` is not meant to be user-facing on the OSS side, we
are going to use it to test some internal and pytorch-owned projects
first. This is not really a great self-serve API for users because it is not
always possible to generate a test given an existing test, we'd only
recommend it to people who know what they're doing.

Due to some co-dev limitations, I'm planning on landing this PR first
and then using it to add crossref testing for internal tests and
fbgemms. I could squash this PR with the internal changes if we want to
see how that works out, just let me know.

Test Plan:
- We create a mini op test suite called MiniOpTests.
- Then, we use `generate_opcheck_tests` to generate tests onto it.
- We have our own test xfail list to check that the things that should
fail do fail.
- Finally, there is a separate TestGenerateOpcheckTests that checks that
the correct number of tests were generated and also tests some helper
functions.

[ghstack-poisoned]

zou3519 commented on torch/testing/_internal/optests/generate_tests.py:
```diff
@@ -0,0 +1,307 @@
+ALL_TESTS = {
+    "test_schema": safe_schema_check,
+    "test_autograd_registration": safe_autograd_registration_check,
+    "test_faketensor": safe_fake_check,
```
Should we name this `test_abstract` to make it more inline with the abstract_impl stuff?

bdhirsh replied:
`test_abstract` to keep inline with the user API's for custom ops sound good to me (but I am only very marginally opinionated)

zou3519 commented on torch/testing/_internal/optests/generate_tests.py:
```diff
@@ -0,0 +1,307 @@
+    The generated test will have name ``test_<something>__<original name>``.
+    For example, if there is a method named ``test_cumsum``, then
+    we will generate a ``test_schema__test_cumsum``,
+    ``test_faketensor__test_cumsum``, etc.
```
Bikeshedding wanted: `test_faketensor__test_cumsum` or `test_cumsum__test_faketensor` or something else?

zou3519 commented on test/test_custom_ops.py:
```diff
@@ -1534,6 +1534,173 @@ def foo_impl(x):
+        op = self._op_delayed_backward_error("delayed_error")
+        x = torch.randn([])
+        y = op(x)
+
+
+mini_op_test_failures_dict = {
+    "MiniOpTest::delayed_error": {
+        "test_aot_dispatch_dynamic__test_delayed_error": "xfail",
+        "test_aot_dispatch_static__test_delayed_error": "xfail",
+    },
+    "MiniOpTest::incorrect_schema": {
+        # "skip" just to test the skip mechanism
+        "test_schema__test_incorrect_schema": "skip",
+    },
+    "MiniOpTest::no_abstract": {
+        "test_aot_dispatch_dynamic__test_no_abstract": "xfail",
+        "test_aot_dispatch_static__test_no_abstract": "xfail",
+        "test_faketensor__test_no_abstract": "xfail",
+    },
+    "aten::nonzero": {
+        # Nonzero doesn't support static shapes
+        "test_aot_dispatch_static__test_nonzero": "xfail",
+    },
+}
+
+mini_op_test_checks = [
+    "test_schema",
+    "test_autograd_registration",
+    "test_faketensor",
+    "test_aot_dispatch_static",
+    "test_aot_dispatch_dynamic",
+]
+
+optests.generate_opcheck_tests(
+    MiniOpTest,
+    ["aten", "MiniOpTest"],
+    mini_op_test_failures_dict,
+    [],
+    mini_op_test_checks,
+)
```
This is an example of how to use the `generate_opcheck_tests` API, including how the failures dict will look like. Please bikeshed.

ezyang commented on torch/testing/_internal/optests/generate_tests.py:
```diff
@@ -0,0 +1,307 @@
+def generate_opcheck_tests(
+    testcase, namespaces, failures_dict, additional_decorators, tests
+):
+    """Given an existing TestCase, use the existing tests to generate
```
Is there any required structure on the TestCase you pass in here?

zou3519 replied:
It should be a subclass of unittest.TestCase. Other than that, no

bdhirsh replied:
Just curious, will the Official(tm) custom ops docs talk about testing? / do you think this will/should eventually be public API?

zou3519 commented on test/test_custom_ops.py:
```diff
@@ -1534,6 +1534,173 @@ def foo_impl(x):
+
+
+mini_op_test_failures_dict = {
+    "MiniOpTest::delayed_error": {
```
TODO(rzou): don't camelcase MiniOpTest: it is a namespace!

zou3519 replied:
TODO(rzou): high-level details into the comment

zou3519 replied:
Question from Ed: how to autogenerate the xfail list?

when to skip vs xfail?
- Legit bugs that segfault

- assertRaisesRegex might be different

Surpression: should trigger fail/xfail at the end of the test. Should probably swallow exceptions and report failure at end of TorchFunctionMode.

zou3519 committed:
Update on "Add `generate_opcheck_tests`, a PT2 crossref testing mechanism"


This PR adds `generate_opcheck_tests`. This is a utility that adds
additional crossref tests to an existing TestCase that has tests that
invokes operators. The main use case is if you have a large test suite
that already exercises operators and want to add automated testing that
the operators are correct, without actually refactoring your code into
something like OpInfos.

Given a `test_` method of a TestCase, we will generate one new
additional test for each of {schema correctness, autograd registration,
faketensor rule, aot_autograd static shapes, aot_autograd dynamic
shapes}. Each newly generated test runs the original test method under a
special torch_function mode (OpCheckMode) that intercepts
`op(*args, **kwargs)` calls and additional passes (op, args, kwargs) to
a separate function (e.g. SchemaCheck).

Nitty-gritty details:
- If a test is named test_cumsum, we end up generating new tests
(`test_schema__test_cumsum`, `test_<something>__test_cumsum`)
- Users can provide a dictionary of expected failures / skips  that is indexed on
operators. This gives us a sense of what operators support PT2 and which
operators require fixing before they support PT2.

Due to some co-dev limitations, I'm planning on landing this PR first
and then using it to add crossref testing for internal tests and
fbgemms. I could squash this PR with the internal changes if we want to
see how that works out, just let me know.

Test Plan:
- We create a mini op test suite called MiniOpTests.
- Then, we use `generate_opcheck_tests` to generate tests onto it.
- We have our own test xfail list to check that the things that should
fail do fail.
- Finally, there is a separate TestGenerateOpcheckTests that checks that
the correct number of tests were generated and also tests some helper
functions.

[ghstack-poisoned]

zou3519 commented on test/test_custom_ops.py:
```diff
@@ -1534,6 +1534,179 @@ def foo_impl(x):
+    "test_aot_dispatch_dynamic",
+]
+
+optests.generate_opcheck_tests(
```
Help wanted: There are two testing designs we can do.
1. This PR: we activate cross-ref testing on a TestCase.
2. We manually run the TestCase to grab sample inputs and generate a Python file that has the sample inputs in it. Then, we set up some OpInfo-style testing that ingests the generated sample inputs. The Python file that has the sample inputs will get checked into the repository (instead of being autogenerated).

This PR goes the first approach, but I really wish we could make the second approach actually work

zou3519 commented on test/test_custom_ops.py:
```diff
@@ -1534,6 +1534,179 @@ def foo_impl(x):
+        y = op(x)
+
+
+mini_op_test_failures_dict = {
```
Help wanted: I don't think it is technically feasible to develop a tool that autogenerates this list. Some tests may need to be marked as "skip" if:
- a test might crash (e.g. the FakeTensor test)
- a test might be flaky (due to randomness, or bugs in PyTorch subsystems)
but it is difficult to detect the above situations.

ezyang replied:
This would be so much easier if we were in OSS lol

ezyang replied:
If you want to go all in on fbcode, xfail generation is automatically done by testx. So you technically don't have to do anything, except when people want to perma-suppress a test because it is definitely not a bug and just a framework limitation that can't be worked around

zou3519 replied:
Testx only does xfails per-test, while each test here may test one or more operators

ezyang commented on torch/testing/_internal/optests/generate_tests.py:
```diff
@@ -0,0 +1,318 @@
+                )
+
+
+class OpCheckMode(TorchFunctionMode):
```
On consequence of having every test be a separate torch function mode is if your test originally took T time, now it takes T * K time, where K is the number of function modes we need to iterate through. It doesn't need to take this long though: all you're really doing is using the pre-existing test code to get real input values to run the operators with, so if memory usage were no concern you could hypothetically just run through the tests once, accurately cloning all of the inputs, and then use these inputs to run through the rest of the tests. You could even use this to do some deduplication (since none of your torch function modes are data dependent, you can keep only one version for any given size that matches).

Granted, this is somewhat difficult to engineer because it's not very natural to have a test suite that runs a bunch of tests first, and THEN runs the rest of the tests. But I feel like there may be ways around this? For example, if you're OK with only running all the tests twice (still an improvement), you can run the tests normally, and then run the tests to setup a fixture, representing all of the test inputs. Then each successive test case picks out what particular input they want to test. Or you could even just stack all of the torch function modes at the same time.

I'm somewhat sensitive to this because when I was working with cross-ref tests I was running them on the PyTorch test suite, which takes a quite long time to run. Maybe all of the internal tests you are working with are quite quick to run; this is less of a concern in that case. (Deduplication was especially a problem with cross-ref tests, since ops were continuously reused to drive other tests). But it is not hard to imagine someone writing a fairly comprehensive test suite and this mechanism quintupling the runtime of those tests. This is not really a good incentive.

zou3519 replied:
> On consequence of having every test be a separate torch function mode is if your test originally took T time, now it takes T * K time, where K is the number of function modes we need to iterate through.

Agree

>  and then run the tests to setup a fixture, representing all of the test inputs. Then each successive test case picks out what particular input they want to test. 

I thought about doing this, but I had trouble engineering this:
- If we run all of the tests a second time before auto-generating tests, then that means it takes the length of the test suite run (20-40 minutes in some cases) before you are actually able to run the autogenerated tests.
- During the second run, we could run the tests one at a time, e.g. test_mm, and then generate additional tests from it. This seems reasonable (and doable in OSS), but it is a bit difficult to dynamically generate tests internally

> Or you could even just stack all of the torch function modes at the same time.

If I am understanding this correctly, your suggestion is when we run `test_mm` the second time, we run all of our tests (schema_check, aot_autograd, ...) during that run. This means that the second `test_mm` run takes a long time and there is no mechanism to separate out what tests we're running (unless we build something to control that).

ezyang commented on torch/testing/_internal/optests/generate_tests.py:
```diff
@@ -0,0 +1,318 @@
+
+def deepcopy_tensors(args, kwargs):
+    def deepcopy_tensor(x):
+        return x.detach().clone().requires_grad_(x.requires_grad)
```
Nyeh, this is not as good as some of the utils in dynamo. In particular this won't preserve striding. Granted, most FB custom ops don't deal with striding, but we should ourselves to a higher standard.

ezyang commented on torch/testing/_internal/optests/generate_tests.py:
```diff
@@ -0,0 +1,318 @@
+
+    # aot_autograd_check runs func(*args, **kwargs) multiple times
+    # and assumes `func` does not modify its inputs.
+    return aot_autograd_check(func, args, kwargs, dynamic, check_gradients="auto")
```
This one is interesting...

ezyang reviewed (approved):
I'm stamping this because I don't think this test mechanism needs to be some perfect, gleaming thing, esp since the idea is for it to be very low effort to add on. The most important future proofing is the xfail strategy, which is worth doing carefully now to minimize work later.

zou3519 committed:
Update on "Add `generate_opcheck_tests`, a PT2 crossref testing mechanism"


This PR adds `generate_opcheck_tests`. This is a utility that adds
additional crossref tests to an existing TestCase that has tests that
invokes operators. The main use case is if you have a large test suite
that already exercises operators and want to add automated testing that
the operators are correct, without actually refactoring your code into
something like OpInfos.

Given a `test_` method of a TestCase, we will generate one new
additional test for each of {schema correctness, autograd registration,
faketensor rule, aot_autograd static shapes, aot_autograd dynamic
shapes}. Each newly generated test runs the original test method under a
special torch_function mode (OpCheckMode) that intercepts
`op(*args, **kwargs)` calls and additional passes (op, args, kwargs) to
a separate function (e.g. SchemaCheck).

Nitty-gritty details:
- If a test is named test_cumsum, we end up generating new tests
(`test_schema__test_cumsum`, `test_<something>__test_cumsum`)
- Users can provide a dictionary of expected failures / skips  that is indexed on
operators. This gives us a sense of what operators support PT2 and which
operators require fixing before they support PT2.

Due to some co-dev limitations, I'm planning on landing this PR first
and then using it to add crossref testing for internal tests and
fbgemms. I could squash this PR with the internal changes if we want to
see how that works out, just let me know.

Test Plan:
- We create a mini op test suite called MiniOpTests.
- Then, we use `generate_opcheck_tests` to generate tests onto it.
- We have our own test xfail list to check that the things that should
fail do fail.
- Finally, there is a separate TestGenerateOpcheckTests that checks that
the correct number of tests were generated and also tests some helper
functions.

[ghstack-poisoned]

zou3519 committed:
Update on "Add `generate_opcheck_tests`, a PT2 crossref testing mechanism"


This PR adds `generate_opcheck_tests`. This is a utility that adds
additional crossref tests to an existing TestCase that has tests that
invokes operators. The main use case is if you have a large test suite
that already exercises operators and want to add automated testing that
the operators are correct, without actually refactoring your code into
something like OpInfos.

Given a `test_` method of a TestCase, we will generate one new
additional test for each of {schema correctness, autograd registration,
faketensor rule, aot_autograd static shapes, aot_autograd dynamic
shapes}. Each newly generated test runs the original test method under a
special torch_function mode (OpCheckMode) that intercepts
`op(*args, **kwargs)` calls and additional passes (op, args, kwargs) to
a separate function (e.g. SchemaCheck).

Nitty-gritty details:
- If a test is named test_cumsum, we end up generating new tests
(`test_schema__test_cumsum`, `test_<something>__test_cumsum`)
- Users can provide a dictionary of expected failures / skips  that is indexed on
operators. This gives us a sense of what operators support PT2 and which
operators require fixing before they support PT2.

Due to some co-dev limitations, I'm planning on landing this PR first
and then using it to add crossref testing for internal tests and
fbgemms. I could squash this PR with the internal changes if we want to
see how that works out, just let me know.

Test Plan:
- We create a mini op test suite called MiniOpTests.
- Then, we use `generate_opcheck_tests` to generate tests onto it.
- We have our own test xfail list to check that the things that should
fail do fail.
- Finally, there is a separate TestGenerateOpcheckTests that checks that
the correct number of tests were generated and also tests some helper
functions.

[ghstack-poisoned]

bdhirsh commented on torch/testing/_internal/optests/generate_tests.py:
```diff
@@ -0,0 +1,319 @@
+
+def safe_aot_autograd_check(op, args, kwargs, dynamic):
+    def func(*args, **kwargs):
+        args, kwargs = pytree.tree_map_only(torch.Tensor, torch.clone, (args, kwargs))
```
If there are any tests internally that rely on sample inputs that are e.g. non-contiguous or autograd-leaves, I think we'll lose that with the `clone()` call here?

Maybe the autograd-leaf metadata is unnecessary, but I think primtorch has a `clone_preserve_strides()` util - that might be marginally safer?

zou3519 replied:
I agree. In the interest of getting this onto the diff train asap I'm going to merge this first and take a follow-up item.

`clone_preserve_strides` will probably be a bit tricky because it calls as_strided and autograd support on that is weird, I think what we really want is aot_autograd_check to clone the inputs respecting strides and autograd leaves each time the forward is ran.

zou3519 commented:
@pytorchbot merge -f "unrelated failures"
